File: app/src/main/java/net/mbonnin/arcanetracker/adapter/ItemAdapter.java
package net.mbonnin.arcanetracker.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import net.mbonnin.arcanetracker.CardDb;
import net.mbonnin.arcanetracker.Deck;
import net.mbonnin.arcanetracker.R;
import net.mbonnin.arcanetracker.Typefaces;
import net.mbonnin.arcanetracker.Utils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;

import timber.log.Timber;

/**
 * Created by martin on 10/17/16.
 */

public class ItemAdapter extends RecyclerView.Adapter {
    static final int TYPE_DECK_ENTRY = 0;
    static final int TYPE_STRING = 1;
    static final int TYPE_HEADER = 2;

    ArrayList<Object> list = new ArrayList<>();

    public void setList(ArrayList list) {
        this.list = list;
        notifyDataSetChanged();
    }

    @Override
    public int getItemViewType(int position) {
        Object o = list.get(position);
        if (o instanceof DeckEntryItem) {
            return TYPE_DECK_ENTRY;
        } else if (o instanceof String) {
            return TYPE_STRING;
        } else if (o instanceof HeaderItem) {
            return TYPE_HEADER;
        }

        Timber.e("unsupported type at position %d: %s", position, o.toString());
        return -1;
    }



    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        Context context = parent.getContext();
        View view;
        switch (viewType) {
            case TYPE_DECK_ENTRY: {
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.bar_card, null);
                break;
            }
            case TYPE_STRING: {
                view = LayoutInflater.from(parent.getContext()).inflate(R.layout.bar_text, null);
                break;
            }
            case TYPE_HEADER: {
                view = LayoutInflater.from(context).inflate(R.layout.bar_header, null);
                break;
            }
            default:
                return null;
        }

        ViewGroup barTemplate = (ViewGroup) LayoutInflater.from(context).inflate(R.layout.bar_template, null);
        RecyclerView.LayoutParams params2 = new RecyclerView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, Utils.dpToPx(30));
        barTemplate.setLayoutParams(params2);
        ViewGroup.LayoutParams params = new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);

        barTemplate.addView(view, 0, params);

        switch (viewType) {
            case TYPE_DECK_ENTRY:
                return new DeckEntryHolder(barTemplate);
            default:
                return new RecyclerView.ViewHolder(barTemplate) {};
        }
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder holder, int position) {
        Object o = list.get(position);
        if (o instanceof DeckEntryItem) {
            DeckEntryItem entry = (DeckEntryItem) o;
            ((DeckEntryHolder)holder).bind(entry);
        } else if (o instanceof String) {
            ViewGroup barTemplate = (ViewGroup) holder.itemView;
            ((TextView)barTemplate.getChildAt(0)).setText((String) o);
        } else if (o instanceof HeaderItem) {
            HeaderItem headerItem = (HeaderItem)o;
            TextView textView = (TextView)holder.itemView.findViewById(R.id.textView);
            textView.setTypeface(Typefaces.belwe());
            String text = "";//headerItem.expanded ?"▼":"▶";
            textView.setText(text + headerItem.title);
            if (headerItem.onClicked != null) {
                textView.setOnClickListener(v -> {
                    headerItem.onClicked.run();
                });
            }
        }
    }

    @Override
    public int getItemCount() {
        return list.size();
    }


}

File: app/src/main/kotlin/net/mbonnin/arcanetracker/ui/overlay/adapter/ItemAdapter.kt
package net.mbonnin.arcanetracker.ui.overlay.adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView
import net.hearthsim.hslog.DeckEntry
import net.mbonnin.arcanetracker.ArcaneTrackerApplication
import net.mbonnin.arcanetracker.R
import net.mbonnin.arcanetracker.Utils

/**
 * Created by martin on 10/17/16.
 */

class ItemAdapter : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    internal var list: List<DeckEntry> = emptyList()

    fun setList(list: List<DeckEntry>) {
        this.list = list
        notifyDataSetChanged()
    }

    override fun getItemViewType(position: Int): Int {
        val o = list[position]
        return when (o) {
            is DeckEntry.Hero -> TYPE_HERO
            is DeckEntry.Item -> TYPE_DECK_ENTRY
            is DeckEntry.Unknown -> TYPE_STRING
            else -> TYPE_HEADER
        }
    }


    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val context = parent.context
        val view: View
        when (viewType) {
            TYPE_HERO -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.bar_hero, null)
            }
            TYPE_DECK_ENTRY -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.bar_card, null)
            }
            TYPE_STRING -> {
                view = LayoutInflater.from(parent.context).inflate(R.layout.bar_text, null)
            }
            TYPE_HEADER -> {
                view = LayoutInflater.from(context).inflate(R.layout.bar_header, null)
            }
            else -> throw Exception("impossibleu!")
        }

        val barTemplate = LayoutInflater.from(context).inflate(R.layout.bar_template, null) as ViewGroup
        val params = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT)

        barTemplate.addView(view, 0, params)

        when (viewType) {
            TYPE_DECK_ENTRY -> return DeckEntryHolder(barTemplate)
            TYPE_HERO -> return HeroHolder(barTemplate)
            else -> return object : RecyclerView.ViewHolder(barTemplate) {

            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val o = list[position]
        when (o) {
            is DeckEntry.Hero -> {
                (holder as HeroHolder).bind(o)
            }
            is DeckEntry.Item -> {
                (holder as DeckEntryHolder).bind(o)
            }
            is DeckEntry.Unknown -> {
                val barTemplate = holder.itemView as ViewGroup
                (barTemplate.getChildAt(0) as TextView).text = ArcaneTrackerApplication.context.getString(R.string.unknown_cards, o.count)
            }
            else -> {
                val textView = holder.itemView.findViewById<TextView>(R.id.textView)
                textView.typeface = ResourcesCompat.getFont(ArcaneTrackerApplication.context, R.font.chunkfive)
                textView.text = when (o) {
                    is DeckEntry.PlayerDeck -> Utils.getString(R.string.deck)
                    is DeckEntry.Secrets -> Utils.getString(R.string.secrets)
                    is DeckEntry.OpponentDeck -> Utils.getString(R.string.allCards)
                    is DeckEntry.Hand -> "${Utils.getString(R.string.hand)} (${o.count})"
                    is DeckEntry.Text -> o.text
                    else -> ""
                }
            }
        }

        if (o is DeckEntry.Text) {
            holder.itemView.setOnClickListener {
                o.onClick?.invoke()
            }
        }
        val params2 = RecyclerView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, Utils.dpToPx(30))
        holder.itemView.layoutParams = params2
    }

    override fun getItemCount(): Int {
        return list.size
    }

    companion object {
        internal val TYPE_DECK_ENTRY = 0
        internal val TYPE_STRING = 1
        internal val TYPE_HEADER = 2
        internal val TYPE_HERO = 3
    }
}
